modules/authentication/src/handlers/tokenProvider.ts

```diff
@@ -0,0 +1,234 @@
+import ConduitGrpcSdk, { Query } from '@conduitplatform/grpc-sdk';
+import { AccessToken, RefreshToken, User } from '../models';
+import { ISignTokenOptions } from '../interfaces/ISignTokenOptions';
+import moment from 'moment/moment';
+import { AuthUtils } from '../utils/auth';
+import * as jwt from 'jsonwebtoken';
+import { Config } from '../config';
+import { Cookie } from '../interfaces/Cookie';
+import { isNil } from 'lodash';
+
+export interface TokenOptions {
+  user: User;
+  clientId: string;
+  config: Config;
+  twoFaPass?: boolean;
+  isRefresh?: boolean;
+  noSudo?: boolean;
+}
+
+export class TokenProvider {
+  private static _instance: TokenProvider;
+
+  private constructor(private readonly grpcSdk: ConduitGrpcSdk) {}
+
+  public static getInstance(grpcSdk?: ConduitGrpcSdk) {
+    if (TokenProvider._instance) return TokenProvider._instance;
+    if (!grpcSdk) throw new Error('GrpcSdk not provided');
+    TokenProvider._instance = new TokenProvider(grpcSdk);
+  }
+
+  async provideUserTokens(tokenOptions: TokenOptions, redirectUrl?: string) {
+    await this.signInClientOperations(
+      this.grpcSdk,
+      tokenOptions.config.clients,
+      tokenOptions.user._id,
+      tokenOptions.clientId,
+    );
+    const [accessToken, refreshToken] = await this.createUserTokens(tokenOptions);
+    const cookies: { accessToken?: Cookie; refreshToken?: Cookie } =
+      this.constructCookies(tokenOptions, [accessToken, refreshToken]);
+    if (Object.keys(cookies).length > 0) {
+      if (redirectUrl) {
+        const redirectUrlWithParams = new URL(redirectUrl);
+        if (!cookies.accessToken) {
+          redirectUrlWithParams.searchParams.append('accessToken', accessToken.token);
+        }
+        if (!cookies.refreshToken && refreshToken) {
+          redirectUrlWithParams.searchParams.append('refreshToken', refreshToken!.token);
+        }
+        return {
+          redirectUrl: redirectUrlWithParams.toString(),
+          setCookies: cookies,
+        };
+      } else {
+        return {
+          result: {
+            message: 'Successfully authenticated',
+            userId: tokenOptions.user._id.toString(),
+            accessToken: cookies.accessToken ?? undefined,
+            refreshToken: cookies.refreshToken ? undefined : refreshToken?.token,
```

Comment:
This doesn't seem right, guessing by the above, it's probably `cookies.refreshToken ?? refreshToken!.token` ?

Reply:
I realized that `refreshToken` may not be set here, but why return `undefined` on a set `cookies.refreshtoken`.
Shouldn't that be `cookies.refreshToken ?? refreshToken?.token` so that it's either the one from cookies or the refreshToken or none instead?